File: density.py
#Load the numpy, math and scipy libraries
import numpy as np
from math import factorial
from scipy.spatial import Delaunay

# Area of a triangle
def volume(sim, points):
    return abs(np.linalg.det(points[sim[1:]] - points[sim[0]])) / factorial(points.shape[1])

# Delaunay tesselation field estimator
class DTFE:
    def __init__(self, points, velocities, m):
        # The Delaunay tesselation
        self.dim = points.shape[1]
        self.delaunay = Delaunay(points)
        self.velocities = velocities
        
        # The density estimate
        self.rho = np.zeros(len(points))
        for sim in self.delaunay.simplices:
            vol = volume(sim, points)
            for index in sim:
                self.rho[index] += vol
        self.rho = (self.dim + 1) * m / self.rho
        
        # The gradients
        self.Drho = np.zeros([len(self.delaunay.simplices), self.dim])
        self.Dv   = np.zeros([len(self.delaunay.simplices), self.dim, self.dim])
        
        for i in np.arange(len(self.delaunay.simplices)):
            sim = self.delaunay.simplices[i]
            p = points[sim] 
            r = self.rho[sim]
            v = velocities[sim]
            A = p[1:] - p[0]
            self.Drho[i] = np.linalg.inv(A) @ (r[1:] - r[0])
            self.Dv[i] = np.linalg.inv(A) @ (v[1:] - v[0])
    
    def density(self, p):
        simplexIndex = self.delaunay.find_simplex(p)
        pointIndex   = self.delaunay.simplices[simplexIndex][0]
        return  self.rho[pointIndex] + self.Drho[simplexIndex] @ (p - self.delaunay.points[pointIndex])
    
    def v(self, p):
        simplexIndex = self.delaunay.find_simplex(p)
        pointIndex   = self.delaunay.simplices[simplexIndex][0]
        return  self.velocities[pointIndex] + self.Dv[simplexIndex] @ (p - self.delaunay.points[pointIndex])

# Bounding Volume Hierarchy
class BoundingVolumeHierarchy:
    def __init__(self, data, box, X, simplices, depth):
        self.left = None
        self.right = None
        self.data = data
        self.box = box
        self.depth = depth
        
        if self.depth != 0:
            self.branch(depth % len(self.box), X, simplices, self.depth - 1)
  
    def branch(self, dim, X, simplices, depth):
        mins = X[simplices[self.data],...,dim].min(axis=1)
        maxs = X[simplices[self.data],...,dim].max(axis=1)
        
        div = self.box[dim].mean()
        L = mins <= div
        R = maxs >= div

        Lbox = np.copy(self.box)
        Lbox[dim] = np.array([Lbox[dim, 0], div])
        Rbox = np.copy(self.box)
        Rbox[dim] = np.array([div, Rbox[dim, 1]])

        data = np.copy(self.data)
        self.data = None
        self.left  = BoundingVolumeHierarchy(data[L], Lbox, X, simplices, depth)
        self.right = BoundingVolumeHierarchy(data[R], Rbox, X, simplices, depth)

    def findCandidateSimplices(self, p):
        dim = self.depth % len(self.box)
        if self.depth == 0:
            return self.data
        elif p[dim] < self.left.box[dim, 1]:
            return self.left .findCandidateSimplices(p)
        elif p[dim] > self.left.box[dim, 1]:
            return self.right.findCandidateSimplices(p)
        else:
            return np.union1d(self.left .findCandidateSimplices(p), 
                              self.right.findCandidateSimplices(p))
    
    def findBox(self, p):
        dim = self.depth % len(self.box)
        if self.depth == 0:
            print('box:', self.box)
        elif p[dim] < self.left.box[dim, 1]:
            self.left.findBox(p)
        elif p[dim] > self.left.box[dim, 1]:
            self.right.findBox(p)
        else:
            self.left .findBox(p)
            self.right.findBox(p)
            
    def findIntersections(self, p, points, simplices):
        candidates = self.findCandidateSimplices(p)
        node_ids = simplices[candidates]

        ori = points[node_ids[:,0]]
        vs = points[node_ids[:,1:]]
        inv_mat = np.linalg.inv(vs - np.repeat(ori, node_ids.shape[1] - 1, axis=0).reshape(vs.shape)).T    
        newp = np.einsum('imk,km->ki', inv_mat, p - ori)
        val = np.all(newp >= 0, axis=1) & np.all(newp <= 1, axis=1) & (np.sum(newp, axis=1) <= 1)
        return candidates[np.nonzero(val)[0]]
    
# Phase-space density estimator
class PhaseSpace:
    def __init__(self, points_init, points, m, depth, box):
        self.dim = points.shape[1]
        self.points = points
        self.simplices = Delaunay(points_init).simplices

        self.BVH = BoundingVolumeHierarchy(
            np.arange(len(self.simplices)), 
            box.astype(float), 
            self.points, self.simplices, self.dim * depth)
        
        self.rho = np.zeros(len(self.simplices), dtype='float64')
        for index, sim in enumerate(self.simplices):
            self.rho[index] = m / volume(sim, self.points) / factorial(self.dim)
    
    def numberOfStreams(self, p):
        return len(self.BVH.findIntersections(p, self.points, self.simplices))

    def density(self, p):
        return self.rho[self.BVH.findIntersections(p, self.points, self.simplices)].sum()
    
# Phase-space Delaunay tesselation field estimator
class PhaseSpaceDTFE:
    def __init__(self, points_init, points, velocities, m, depth, box):
        self.dim = points.shape[1]
        self.points = points
        self.velocities = velocities
        self.simplices = Delaunay(points_init).simplices

        self.BVH = BoundingVolumeHierarchy(
            np.arange(len(self.simplices)), 
            box.astype(float), 
            self.points, self.simplices, self.dim * depth)
        
        # The density estimate
        self.rho = np.zeros(len(points), dtype='float64')
        for sim in self.simplices:
            vol = volume(sim, self.points)
            for index in sim:
                self.rho[index] += vol
        self.rho = (self.dim + 1) * m / self.rho
        
        # The gradients
        self.Drho = np.zeros([len(self.simplices), self.dim], dtype='float64')
        self.Dv   = np.zeros([len(self.simplices), self.dim, self.dim])
        
        for i in np.arange(len(self.simplices)):
            sim = self.simplices[i]
            p = points[sim] 
            r = self.rho[sim]
            v = velocities[sim]
            A = p[1:] - p[0]
            self.Drho[i] = np.linalg.inv(A) @ (r[1:] - r[0])
            self.Dv[i] = np.linalg.inv(A) @ (v[1:] - v[0])
    
    def numberOfStreams(self, p):
        return len(self.BVH.findIntersections(p, self.points, self.simplices))

    def density(self, p):
        simplexIndices = self.BVH.findIntersections(p, self.points, self.simplices)
        
        dens = 0
        for simplexIndex in simplexIndices:
            pointIndex = self.simplices[simplexIndex][0]
            dens = dens + self.rho[pointIndex] + self.Drho[simplexIndex] @ (p - self.points[pointIndex])
        return dens
    
    def v(self, p):
        simplexIndices = self.BVH.findIntersections(p, self.points, self.simplices)
        
        vs = np.zeros([len(simplexIndices), 2])
        for i, simplexIndex in enumerate(simplexIndices):
            pointIndex   = self.simplices[simplexIndex][0]
            vs[i] = self.velocities[pointIndex] + self.Dv[simplexIndex] @ (p - self.points[pointIndex])
        return vs

File: density.jl
using LinearAlgebra
using Delaunay

struct BVH
    data::Union{Vector{Int}, Nothing}
    box::Matrix{Float64}
    depth::Int
    left::Union{BVH, Nothing}
    right::Union{BVH, Nothing}
end

BVH_constructor = function(data, box, points, simplices, depth)
    if depth != 0
        dim = (depth % size(box)[1]) + 1
        mins = vec(minimum(positions[simplices[data, :], dim], dims = 2))
        maxs = vec(maximum(positions[simplices[data, :], dim], dims = 2))
        
        mean(x) = sum(x) / length(x)

        div = mean(box[dim,:])
        L = mins .<= div
        R = maxs .>= div

        Lbox = copy(box)
        Lbox[dim,:] = [Lbox[dim, 1], div]
        Rbox = copy(box)
        Rbox[dim,:] = [div, Rbox[dim, 2]]

        BVH(nothing, box, depth, BVH_constructor(data[L], Lbox, points, simplices, depth - 1), 
                                 BVH_constructor(data[R], Rbox, points, simplices, depth - 1))
    else 
        BVH(data, box, depth, nothing, nothing)
    end
end

findCandidateSimplices = function (p::Vector{Float64}, BVH_tree::BVH)
    dim = (BVH_tree.depth % size(BVH_tree.box)[1]) + 1
    if BVH_tree.depth == 0
        return BVH_tree.data
    elseif p[dim] < BVH_tree.left.box[dim, 2]
        return findCandidateSimplices(p, BVH_tree.left)
    elseif p[dim] > BVH_tree.left.box[dim, 2]
        return findCandidateSimplices(p, BVH_tree.right)
    else
        return union(findCandidateSimplices(p, BVH_tree.left),
                     findCandidateSimplices(p, BVH_tree.right))
    end
end

findBox = function (p::Vector{Float64}, BVH_tree::BVH)
    dim = (BVH_tree.depth % size(BVH_tree.box)[1]) + 1
    if BVH_tree.depth == 0
        println(BVH_tree.box)
    elseif p[dim] < BVH_tree.left.box[dim, 2]
        return findBox(p, BVH_tree.left)
    elseif p[dim] > BVH_tree.left.box[dim, 2]
        return findBox(p, BVH_tree.right)
    else
        findBox(p, BVH_tree.left)
        findBox(p, BVH_tree.right)
    end
end

intersection = function (p::Vector{Float64}, simplex::Matrix{Float64})
    barry = inv(simplex[2:end,:]' .- simplex[1,:]) * (p .- simplex[1,:])
    return all(barry .>= 0) & all(barry .<= 1) & (sum(barry) <= 1.)
end

findIntersections = function (p::Vector{Float64}, BVH_tree::BVH, points, simplices)
    candidates = findCandidateSimplices(p, BVH_tree)
    filter(i -> intersection(p, points[simplices[i,:],:]), candidates)
end

volume(sim, points) = abs(det(points[sim[2:end],:]' .- points[sim[1],:])) /  factorial(size(points)[2])

struct ps_dtfe
    rho::Vector{Float64}
    Drho::Matrix{Float64}
    Dv::Array{Float64}
    tree::BVH
    simplices::Matrix{Int}
    positions::Matrix{Float64}
    velocities::Matrix{Float64}
end

ps_dtfe_constructor = function (positions_initial, positions, velocities, m, depth, box)
    simplices = delaunay(positions_initial).simplices
    dim = size(box)[1]

    BVH_tree = BVH_constructor(1:size(simplices)[1], box, positions, simplices, depth * dim)

    rho = zeros(size(positions)[1])
    for i in axes(simplices, 1)
        vol = volume(simplices[i,:], positions)
        for index in simplices[i,:]
            rho[index] += vol
        end
    end
    rho = (1. + dim) * m ./ rho

    Drho = zeros(size(simplices)[1], dim)
    Dv   = zeros(size(simplices)[1], dim, dim)

    for i in axes(simplices, 1)
        p = positions[simplices[i,:],:]
        r = rho[simplices[i,:],:]
        v = velocities[simplices[i,:],:]
        A_inv = inv(p[2:end,:]' .- p[1,:])'
        Drho[i,:] = A_inv * (r[2:end] .- r[1])
        Dv[i,:,:] = A_inv * ((v[2:end,:]' .- v[1,:])')
    end

    return ps_dtfe(rho, Drho, Dv, BVH_tree, simplices, positions, velocities)
end

density = function (p::Vector{Float64}, estimator::ps_dtfe)
    simplexIndices = findIntersections(p, estimator.tree, estimator.positions, estimator.simplices)

    dens = 0.
    for simplexIndex in simplexIndices
        pointIndex = estimator.simplices[simplexIndex,1]
        dens += estimator.rho[pointIndex] + estimator.Drho[simplexIndex,:]' * (p .- estimator.positions[pointIndex,:])
    end
    return dens
end

v = function (p::Vector{Float64}, estimator::ps_dtfe)
    simplexIndices = findIntersections(p, estimator.tree, estimator.positions, estimator.simplices)

    vs = zeros(length(simplexIndices), length(p))

    for (i, simplexIndex) in pairs(simplexIndices)
        pointIndex = estimator.simplices[simplexIndex, 1]
        vs[i,:] = estimator.velocities[pointIndex,:] + estimator.Dv[simplexIndex,:,:] * (p .- estimator.positions[pointIndex,:])
    end
    return vs
    
end

numberOfStreams = function (p::Vector{Float64}, estimator::ps_dtfe)
    simplexIndices = findIntersections(p, estimator.tree, estimator.positions, estimator.simplices)
    return length(simplexIndices)
end